kind.go

```diff
@@ -45,48 +47,204 @@ func (m Maturity) String() string {
+//
+// Directly implementing this interface is discouraged. Strongly prefer instead to
+// rely on [BindCore], [BindCustom], or [BindComposable].
+//
+// [the kindsys repository]: https://github.com/grafana/kindsys
 type Kind interface {
-	// Props returns a [kindsys.SomeKindProps], representing the properties
+	// Name returns the kind's name, as specified in the name field of the kind definition.
+	//
+	// Note that this is the capitalized name of the kind. For other names and
+	// common kind properties, see [Props.CommonProperties].
+	Name() string
+
+	// MachineName returns the kind's machine name, as specified in the machineName
+	// field of the kind definition.
+	MachineName() string
+
+	// Maturity indicates the maturity of this kind, one of the enum of values we
+	// accept in the maturity field of the kind definition.
+	Maturity() Maturity
+
+	// Props returns a [SomeKindProps], representing the properties
 	// of the kind as declared in the .cue source. The underlying type is
 	// determined by the category of kind.
-	//
-	// This method is largely for convenience, as all actual kind categories are
-	// expected to implement one of the other interfaces, each of which contain
-	// a Def() method through which these same properties are accessible.
 	Props() SomeKindProperties
 
-	// TODO docs
-	Lineage() thema.Lineage
+	// CurrentVersion returns the version number of the schema that is considered
+	// the 'current' version, usually the latest version. When initializing object
+	// instances of this Kind, the current version is used by default.
+	CurrentVersion() thema.SyntacticVersion
 
-	// TODO remove, unnecessary with Props()
-	Name() string
+	// Lineage returns the kind's [thema.Lineage]. The lineage contains the full
+	// history of object schemas associated with the kind.
+	//
+	// TODO separate this onto an optional, additional interface
+	Lineage() thema.Lineage
+}
 
-	// TODO remove, unnecessary with Props()
-	MachineName() string
+// ResourceKind represents a kind that defines a root object, or Kubernetish resource.
+//
+// TODO this is a temporary intermediate while we combine [Core] and [Custom]. This name will probably go away.
+type ResourceKind interface {
+	Kind
 
-	// TODO remove, unnecessary with Props()
-	Maturity() Maturity // TODO unclear if we want maturity for raw kinds
+	// Validate takes a []byte representing an object instance of this kind and
+	// checks that it is a valid instance of at least one schema defined in the
+	// kind.
+	//
+	// A decoder must be provided that knows how to decode the []byte into an
+	// intermediate form. At minimum, the right decoder must be chosen for the
+	// format - for example, JSON vs YAML. For resource kinds, a decoder must
+	// also know how to transform the input from a Kubernetes resource object
+	// shape to Grafana's object shape. See [github.com/grafana/kindsys/encoding].
+	Validate(b []byte, codec Decoder) error
```

Comment:
Does it make sense to have a separate `Validate` method when `FromBytes` would in theory do the exact same thing? I guess the only optimization is reduced memory allocation, though it depends on how the interface is implemented.

Reply:
The key difference is that `Validate` only needs to check whether there _is_ a schema that the input is valid against - not actually perform translation. This is a pretty direct mapping to `lineage.ValidateAny()` vs. a vmuxer

Reply:
It does, because `FromBytes` executes translations (to the `CurrentVersion()`), whereas `Validate` merely tries to validate that there's _some_ schema against which the bytes are valid.